helper/salt/salt.go

```diff
@@ -57,46 +73,91 @@ func NewSalt(view logical.Storage, config *Config) (*Salt, error) {
 		config: config,
 	}
 
-	// Look for the salt
-	raw, err := view.Get(config.Location)
-	if err != nil {
-		return nil, fmt.Errorf("failed to read salt: %v", err)
-	}
-
-	// Restore the salt if it exists
-	if raw != nil {
-		s.salt = string(raw.Value)
+	var raw *logical.StorageEntry
+	var err error
+	if config.StaticSalt != "" {
+		s.salt = config.StaticSalt
+	} else {
+		if view != nil {
+			// Look for the salt
+			raw, err = view.Get(config.Location)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read salt: %v", err)
+			}
+
+			// Restore the salt if it exists
+			if raw != nil {
+				s.salt = string(raw.Value)
+			}
+		}
 	}
 
 	// Generate a new salt if necessary
 	if s.salt == "" {
 		s.salt = uuid.GenerateUUID()
 		s.generated = true
-		raw = &logical.StorageEntry{
-			Key:   config.Location,
-			Value: []byte(s.salt),
+		if view != nil {
+			raw = &logical.StorageEntry{
+				Key:   config.Location,
+				Value: []byte(s.salt),
+			}
+			if err := view.Put(raw); err != nil {
+				return nil, fmt.Errorf("failed to persist salt: %v", err)
+			}
+		}
+	}
+
+	if config.HMAC != nil {
+		if len(config.HMACType) == 0 {
+			return nil, fmt.Errorf("HMACType must be defined")
 		}
-		if err := view.Put(raw); err != nil {
-			return nil, fmt.Errorf("failed to persist salt: %v", err)
+		s.hmac = hmac.New(config.HMAC, []byte(s.salt))
+		if s.hmac == nil {
+			return nil, fmt.Errorf("failed to instantiate HMAC function")
 		}
+		s.hmacType = config.HMACType
 	}
+
 	return s, nil
 }
 
-// SaltID is used to apply a salt and hash functio to an ID to make sure
-// it is not reversable
+// SaltID is used to apply a salt and hash function to an ID to make sure
+// it is not reversible
 func (s *Salt) SaltID(id string) string {
 	return SaltID(s.salt, id, s.config.HashFunc)
 }
 
+// GetHMAC is used to apply a salt and hash function to an ID to make sure
+// it is not reversible, with an additional HMAC
+func (s *Salt) GetHMAC(id string) string {
+	if s.hmac == nil {
+		return ""
+	}
+	s.hmac.Reset()
```

Comment:
I see whats happening, yeah this isn't thread safe. Better to allocate new HMAC.